### ISSUE TITLE
basicTouchScreen.ino not working with portenta and TFT V2

### ISSUE THREAD

**Issue by hpssjellis** [Lennart401/PortentaTouchScreen#2]
Your example basicTouchScreen.ino with 

```
// y- x- y+ x+ connected to YD, XL, YU, XR
PortentaTouchScreen ts(5, A6, A4, 4);
```

Which I assume is:   A0,     A1,      A2,     A3        on the TFT V2
connected to:           D5,     A6,      A4,     D4      on  the Portenta

I also connected +5V on Portenta to 5V on TFT and GND to GND.
Strangely Vin on the TFT did not light the screen but 5V does.

Anyway, no results. any suggestions. Even if I have A0,A1,A2,A3 to the wrong values I should still get some rsults?

Any other pins needing connections?

By the way, thanks for the explanation about A0-A3 being different from A4-A6 that might help with a few things I am working on.

**Comment by Lennart401** [comment]
Yes, I am fully aware that the example is not working correctly with that pin setup.

Your assumtion is fully correct:

- YD "y down" maps to A0 on the TFT v2
- XL "x left" maps to A1 on the TFT v2
- YU "y up" maps to A2 on the TFT v2
- XR "x right" maps to A3 on the TFT v2

The letters YD, XL, YU, XR are actually printed on the wide brown connector cable of the TFT: ![touchscreen connector cable](https://user-images.githubusercontent.com/65546425/132014282-8fe02c98-2dda-41e1-afe9-aacffd77a67b.jpg)

So just to clarify, the touchscreen it not connected to the rest of the TFT v2 electronically in any way. It is simply glued onto the screen and its four wires from the brown connector just run over the PCB to the header pins. This means connecting the display part of the TFT v2 is another story that this library does not cover.

However, to make the TFT v2 display part work, there are 8 pins that need to be connected. 

| TFT v2 | Arduino Portenta |
| ------ | ---------------- |
| RESET | RESET |
| 5V | 5V |
| GND (on of them) | GND |
| ------ | ---------------- |
| D6 | Any free digital I/O pin, I used D6 |
| D5 | D7 (SPI1 CS) |
| ------ | ---------------- |
| D11 | D8 (SPI1 MOSI) |
| D12 | D10 (SPI1 MISO) |
| D13 | D9 (SPI1 CK) |

I have used the [Adafruit ILI9341 library](https://github.com/adafruit/Adafruit_ILI9341) to talk to the display.

For connection of the touchscreen part of the TFT, see my answer on [Issue #1](https://github.com/Lennart401/PortentaTouchScreen/issues/1#issuecomment-912543135).

**Comment by hpssjellis** [comment]
@Lennart401 This is so helpful. Thank you.

**Comment by hpssjellis** [comment]
Just testing your basic setup. Touch screen works on it's own, now trying the above with the LCD, do you make any changes to the code? 

https://github.com/adafruit/Adafruit_ILI9341/tree/master/examples/graphicstest


 Thinking these should change [here](https://github.com/adafruit/Adafruit_ILI9341/blob/028250a056241585a096728dd599208291e1c1d3/examples/graphicstest/graphicstest.ino#L22-L23)


```
// For the Adafruit shield, these are the default.
#define TFT_DC 9
#define TFT_CS 10
```

Can you think of any other minor changes to the LCD test?

With the breakout board I will try A7, but I can also try one of the other 2 SPI sets of pins. If I can get that working I think we need to talk to Arduino about if it can be done without the Breakout board. Bit of a shame to need the huge board when the Portenta form factor is so small.

**Comment by Lennart401** [comment]
[This](https://github.com/Lennart401/LernfabrikArduino/blob/main/CoreM4/CoreM4.ino) setup is working for me, it is from my bigger Arduino project. I don't think I did any other modifications to the graphics test except for the DC & CS pins.

_TLDR;_ `TFT_DC` should be `6` and `TFT_CS` should be `7` for my suggested pin setup._

---

Even if you can get that to work (which would be absolutely fantastic, though!), that would still require the breakout board. I absolutely agree with you, it is a bit frustrating that without the breakout board, only one out of 7 analog pins is fully independently usable. Every single Arduino Uno can do better than that!

**Comment by hpssjellis** [comment]
@Lennart401 Can you check my pin allocation. My Basic Adafruit_ILI9341 examples aren't working. 

Image Old, see below.

I am not connecting the pressure pins until I get these examples working.

**Comment by hpssjellis** [comment]
@Lennart401 

The plot thickens:


The following code works using my Arduino Yun and my TFT V2 shield using the seeeduino TFT v2 library. (So easy since it plugs in, arduino needs to make an adaptor board from the  160 pins to the old style.)


```
#include <SPI.h>
#include <stdint.h>
#include <SeeedTouchScreen.h>
#include <TFTv2.h>


#define ShowSerial Serial

#define TS_MINX 116*2
#define TS_MAXX 890*2
#define TS_MINY 83*2
#define TS_MAXY 913*2


// For better pressure precision, we need to know the resistance
// between X+ and X- Use any multimeter to read it
// The 2.8" TFT Touch shield has 300 ohms across the X plate
                             //  both, both, analog, analog
//TouchScreen ts = TouchScreen(XP, YP, XM, YM); 
TouchScreen ts = TouchScreen(A3, A2, A1, A0);

void setup(void) {


    ShowSerial.begin(9600);
    TFT_BL_ON;      // turn on the background light
    Tft.TFTinit();  // init TFT library
    Tft.fillScreen(0, 240, 0, 320, BLUE);

    // no proof this text orientation works!
    TextOrientation orientation;
    orientation = PORTRAIT_VERTICAL; // LANDSCAPE; PORTRAIT_BACKWARDS, PORTRAIT_UPSIDE_DOWN_BACKWARDS,PORTRAIT_UPSIDE_DOWN, PORTRAIT_VERTICAL

}

void loop(void) {
    // a point object holds x y and z(pressure) coordinates
    Point p = ts.getPoint();

    p.x = map(p.x, TS_MINX, TS_MAXX, 0, 240);
    p.y = map(p.y, TS_MINY, TS_MAXY, 0, 320);

    Tft.drawString("Cool" , 90, 140, 3, GREEN);

    // we have some minimum pressure we consider 'valid'
    // pressure of 0 means no pressing!
    
    if (p.z > __PRESSURE) {
   // if (p.z > 0) {
        ShowSerial.println("x: " + String(p.x) + ", y: " + String(p.y) + ", pressure: " + String(p.z));

        int myRadius = (500-p.z)/10;  // format pressure
        Tft.drawCircle(p.x, p.y, myRadius, random(0xFFFF));    //draw circle, center:(119, 160), color: random

    }

    delay(50);  // slow it down for no real reason
}

```


but using the Adafruit_ILI9341 library I get a serial printout but nothing active on the screen other than a grey backdrop.  Any suggestions?

I would use the TFTv2 library since the pressure stuff works, but it does not compile on my portenta. Missing the "pgmspace.h" file

**Comment by hpssjellis** [comment]
@Lennart401 

Did some detective work and connected every pin from the YUN with wires to the TFTV2. It did not work! Connected the 6 prong pins at the bottom of the board and it worked? What the heck are those pins. I thought they would be for debugging.

**Comment by hpssjellis** [comment]
Well after a few revisions this makes some sense now. Still need to test it but this is what works using the TFTV2 on the Yun for me. The Touch Sensor is listed for what works on the Portenta. On the Yun it is an easy A0->A0, A1->A1,A2>A2,A3->A3 unlike the image


![image](https://user-images.githubusercontent.com/5605614/132150621-32d36c14-28a6-46fb-875d-8d5d0089f6e1.png)

**Comment by hpssjellis** [comment]
No worries. Got it working, just have to finish up a few things.

**Comment by hpssjellis** [comment]
So here is my code that works. Strangely unlike on the Yun the edges of the touch screen do not seem very responsive, I have to figure that out. Also tested A5-A0 and only A5 gets pressure readings but cuts off any SPI communication.

Working code needs breakout board for A7 only.

```

// demonstrates drawing and touch operations.
// needs this library  Adafruit_ILI9341
// needs this library install as zip file https://github.com/Lennart401/PortentaTouchScreen
//


#include <Adafruit_GFX.h>
#include <Adafruit_ILI9341.h>
#include <PortentaTouchScreen.h>


// PortentaTouchScreen ts(XP, YP, XM, YM, 300);  // original
// PortentaTouchScreen ts(A3, A2, A1, A0, 300);  // Yun
// Must be:       digital, analog/digital, analog/digital, digital, resistor across x plate ~300 ohms
// Touchscreen          X+, X-, Y+, Y-, resistor 
PortentaTouchScreen ts(D5, A6, A7, D4, 300);  // on Portenta

// Set for Portenta
#define TFT_CS D7
#define TFT_DC D6
Adafruit_ILI9341 tft = Adafruit_ILI9341(TFT_CS, TFT_DC);


void setup(void)
{
  Serial.begin(115200);
  tft.begin();

  tft.fillScreen(ILI9341_BLUE);
  // origin = left,top landscape (USB left upper)
  tft.setRotation(1); 
  
  tft.setTextColor(ILI9341_WHITE);
  tft.setTextSize(2);
  tft.setCursor(100, 20);  //before print set x,y
  tft.println("Working!");

}


void loop(){
  
    // a point object holds x y and z(pressure) coordinates
    Point p = ts.getPoint();

    int myX      = map(p.x(), 200, 1400, 0, 240);   // map raw x to real x
    int myY      = map(p.y(), 600, 1200, 320, 0);   // map raw y to real y
    int myRadius = map(p.z(), 50, 700, 5, 70);      // map raw Pressure to real Radius
    
    // we have some minimum pressure we consider 'valid'
    // pressure of 0 means no pressing!
    
     if (p.z() > 0) {  // minimum presure to register
      
        Serial.println(" raw x: " + String(p.x()) + ",\t  raw y: " + String(p.y()) + ",\t raw pressure: " + String(p.z()));
        Serial.println("real x: " + String(myX)   + ",\t real y: " + String(myY)   + ",\t real  radius: " + String(myRadius));
        Serial.println("");

        tft.drawCircle(myX, myY, myRadius, random(0xFFFF));    //draw circle, center:(119, 160), color: random
    }
    delay(100);  // 100 slow it down to read serial monitor can speed it up when working
}



``` 


![image](https://user-images.githubusercontent.com/5605614/132236228-267b170f-77cd-4560-82f2-d02563e321ed.png)


----

@Lennart401 

  Ever thought of making a high density connector PCB board for the Portenta? This guy is on the Portenta Face book page 
https://www.facebook.com/groups/portenta/?multi_permalinks=838667036787840&notif_id=1630612109568129&notif_t=group_activity&ref=notif

his throne PCB is

https://www.facebook.com/watch/?v=788409832111335

**Comment by Lennart401** [comment]
Wow, that's some great results! Sorry I couldn't reply earlier, I was busy with other stuff.

I'll make an update to the basicTouchScreen example and incorporate your results into the library README.

As soon as I get can my hands on a portenta breakout board, I'll try working with the touchscreen (I did move to rotary encoders as input devices in the meantime) again, maybe I can find the reason why the touchscreen does not seem to be very responsive at the edges. One possible clue could be the fairly high "LOW"-voltages (~0.7V to GND) I occasionally observed on the portenta, but there might be more to it.

Concerning the high density connector PCB for the portenta: first of all, thanks for the input! I might actually design a PCB with high density connectors at some point, probably in the form of "shield" board, i. e. hold the TFT v2 on top and have some extra connectors to the sides. Something like this:
![PCB Design idea](https://user-images.githubusercontent.com/65546425/132342783-13a8680d-0423-4847-b8cf-b86e752521ed.jpg)

I'll have to run some tests using the breakout board first though, plus this will not be my top priority for a while, since I'm still very much in the prototyping phase of my main project and there are other things that need to be finished first. Great idea however.

**Comment by hpssjellis** [comment]
@Lennart401 I know nothing about pcb design, but I think some companies help to make it fairly easy. One thing I would consider is like the breakout board the high desnity connectors can be connected on both sides. That would allow the vision shield to be on the back and the TFTV2 to be on the front. Would be a powerful combination. Sad your project just needs the one A7 connection, seems like a lot of work. 

A more useful pcb would be a portenta to Yun style converter shield. That would be really useful for all kinds of old connections.

**Comment by hpssjellis** [comment]
So I posted this on the portenta facebook page. Any idea to make a single HD connector shield, for situations like your issue with needing A7 connectivity [here](https://www.facebook.com/groups/portenta/posts/841538866500657/?__cft__[0]=AZX6WuD5C8HJljlR6HjvYi3wnozbmNYA0Xq80sFUoVKivTbMq2XTHauuhsN_XkJzBldkMijuIOhZ_8nFHbuExfMYEURdbJU6862d_TOlpjljN6cLYfMfVLzZR4xmgcCkGri46sDWdGGVZOW3z23Qw6iq&__tn__=%2CO%2CP-R).

**Comment by Lennart401** [comment]
I would probably expose all pins I need using the HD connectors just for simplicity and space saving (e.g. I don't have to do any soldering on the portenta itself).

Mirroring the HD connectors to the other side of the PCB should be pretty easy, just place the same type of connector like on the portenta straight under the top-side connectors (which the portenta connects to) and use vias through the PCB to connect the top and bottom layers.

My idea is that the top side of the PCB would be like the Yun/Uno boards, but where the chips are, there would be the two HD connectors the portenta would mount to. Then, you would mirror the HD connectors to the bottom of the PCB, so you could connect e.g. the vision shield. Since the top of side of the PCB would have the same pin layout like the Yun/Uno, you could solder female pin headers to it for connecting all kinds of old-style shields (e.g. the TFT).

I'll have a look at FB page and see what that brings up.

**Comment by hpssjellis** [comment]
Yes, what you want to do Arduino should have done when the Portenta first came out. That would give backward compatibility for a ton of shields. I will try to ask the Arduino people on the facebook page.

**Comment by Lennart401** [comment]
I absolutely agree with you, that would have been a total game-changer. I guess they wanted to create a "new" environment with the portenta and the accompanying "pro"-shields (that clamp to the HD connectors). 

Thank you very much that, we'll see what the future (or Arduino) will bring us.

**Comment by hpssjellis** [comment]
Hi @Lennart401 . I don't think this solves the problem but I never knew about it.    

It is a MKR to UNO adapter. Just thought you should see it


 [https://store-usa.arduino.cc/collections/mkr-family/products/mkr2uno-adapter]( https://store-usa.arduino.cc/collections/mkr-family/products/mkr2uno-adapter)

**Comment by Lennart401** [comment]
No, I doesn't solve the problem, but that's like the board I imagined. Just with the portenta HD connectors instead of the traditional pin headers (and a few added features like the HD connectors on the back of the board, for e.g. the vision shield).

Thanks for pointing it out though, I never knew about it either.

**Comment by hpssjellis** [comment]
I always have my Portenta with female headders coming up from the board. Too bad the above board doesn't have male pins going down, then the Portenta could plug right in to the above board from below.
